src/drivers/can.rs

```diff
@@ -0,0 +1,1502 @@
+
+        /* Sanity check, reading so the configuration has written correctly */
+        let canctrl_bits = self.read_register(MCP2515Register::CANCTRL, 0x00).unwrap();
+        let txrtscttrl_reg = self
+            .read_register(MCP2515Register::TXRTSCTRL, 0x00)
+            .unwrap();
+        let bfpctrl_reg = self.read_register(MCP2515Register::BFPCTRL, 0x00).unwrap();
+
+        /* The requested mode must be varified by reading the OPMOD[2:0] bits
+         * (CANSTAT[7:5]) */
+        let canstat_new = self.read_register(MCP2515Register::CANSTAT, 0x00).unwrap();
+        defmt::println!("CANSTAT after reset instruction is: {:08b}", canstat_reg);
+        defmt::println!("CANCTRL mask prior writing: {:08b}", canctrl_byte);
+        defmt::println!("CANCTRL after setup is: {:08b}", canctrl_bits);
+        defmt::println!("CANSTAT after setup is: {:08b}", canstat_new);
+        defmt::println!("TXRTSCTRL after setup is: {:08b}", txrtscttrl_reg);
+        defmt::println!("BFPCTRL after setup is: {:08b}", bfpctrl_reg);
+
+        self
+    }
+
+    ///This change the mode and set it to Normal mode for sending and receiving
+    /// on the CAN bus.
+    pub fn activate_canbus(&mut self) {
+        const ONE_SHOT_MODE: bool = true;
+        if (self.can_settings.canctrl.mode != OperationTypes::Normal) {
+            //let mut can_settings = CanSettings::default();
+            let canctrl_settings = SettingsCanCtrl::new(
+                OperationTypes::Normal,
+                true,
+                CLKPRE::DIV1,
+                false,
+                !ONE_SHOT_MODE,
+            );
+            self.can_settings.canctrl = canctrl_settings;
+            defmt::info!("Setting Operation Type to: Normal mode");
+            self.change_settings(self.can_settings);
+        }
+    }
+
+    /// Transmit frame on CAN bus, by loading TXBN buffer, and write
+    /// 'Request-To-Send' instruction.
+    ///
+    /// ## TODO
+    ///
+    /// Add logic for targeting TX buffers to read, that is not busy...
+    fn transmit_can(&mut self, can_msg: &CanMessage) {
+        let can_frame = CanMessage::new(can_msg.id, &can_msg.data).unwrap();
+        self.load_tx_buffer(TXBN::TXB0, can_frame); // How should I handle buffer target for transmission
+        self.request_to_send(TXBN::TXB0);
+    }
+
+    /// Receive CAN frame by reading the appropriate RXBN buffer.
+    ///
+    /// ## TODO
+    ///
+    /// Write logic for picking appropriate RXBN buffer to read.
+    /// Based on CANINTF, RXNCTRL.FILT bits, or through RX status instruction.
+    fn receive_can(&mut self) -> CanMessage {
+        let rx_buffer = self.read_rx_buffer(RXBN::RXB0).unwrap();
+        let mut frame = CanMessage::try_from(rx_buffer).unwrap();
+        defmt::info!("Can bus received the Frame:");
+        frame.print_frame();
+        let data_str: &str = core::str::from_utf8(frame.data()).unwrap();
+        defmt::info!("Decoded frame data: {:?}", data_str);
```

Comment:
Libraries should typically not write to the equivalent of stdout, I think that we should remove the defmt dependencies. If not now we should do it once we break it out in to a distinct crate

Reply:
Will do this when we break it out in its own crate!